contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol

```diff
@@ -159,7 +159,9 @@ contract OVM_ExecutionManager is iOVM_ExecutionManager, Lib_AddressResolver {
         override
         public
     {
-        require(transactionContext.ovmNUMBER == 0, "Only be callable at the start of a transaction");
+        require(transactionContext.ovmNUMBER == 0, "Only callable at the start of a transaction");
+        require(_transaction.blockNumber != 0, "Prevent reentrancy by disallowing default context values");
```

Comment:
The [issue suggests](https://github.com/ethereum-optimism/roadmap/issues/803#issuecomment-801442414) two options: 

> 1. in run, ban the transaction.blocknumber argument from being passed as its default value, so that you cannot ever get through run in a way which is re-enterable
> 2. add an explicit reentrancy guard variable on run

I initially preferred to add a `nonreentrant` modifier, as it's straightforward and easy to follow. However given that it added  >5000 gas, I decided against this. 

This approach is IMO less readable, but does the trick.